test-data/unit/check-overloading.test

```diff
@@ -1723,6 +1782,127 @@ def foo2(**kwargs: int) -> str: ...
+reveal_type(foo(*[1, 2]))     # E: Revealed type is '__main__.C'
+reveal_type(foo(*[1, 2, 3]))  # E: Revealed type is '__main__.C'
+
+x: List[int]
+reveal_type(foo(*x))  # E: Revealed type is '__main__.C'
+
+y: List[str]
+foo(*y)  # E: No overload variant of "foo" matches argument type "List[str]"
+[builtins fixtures/list.pyi]
+
+[case testOverloadMultipleVarargDefinition]
+from typing import overload, List, Any
+
+class A: ...
+class B: ...
+class C: ...
+class D: ...
+
+@overload
+def foo(x: int) -> A: ...
+@overload
+def foo(x: int, y: int) -> B: ...
+@overload
+def foo(x: int, y: int, z: int, *args: int) -> C: ...
+@overload
+def foo(*x: str) -> D: ...
+def foo(*args): pass
+
+reveal_type(foo(*[1, 2]))      # E: Revealed type is '__main__.C'
+reveal_type(foo(*["a", "b"]))  # E: Revealed type is '__main__.D'
+
+x: List[Any]
+reveal_type(foo(*x))  # E: Revealed type is 'Any'
+[builtins fixtures/list.pyi]
+
+[case testOverloadMultipleVarargDefinitionComplex]
+from typing import TypeVar, overload, Any, Callable
+
+T1 = TypeVar('T1')
+T2 = TypeVar('T2')
+T3 = TypeVar('T3')
+
+@overload
+def chain_call(input_value: T1,
+               f1: Callable[[T1], T2]) -> T2: ...
+@overload
+def chain_call(input_value: T1,
+               f1: Callable[[T1], T2],
+               f2: Callable[[T2], T3]) -> T3: ...
+@overload
+def chain_call(input_value: T1,
+               *f_rest: Callable[[T1], T1]) -> T1: ...
+@overload
+def chain_call(input_value: T1,
+               f1: Callable[[T1], T2],
+               f2: Callable[[T2], T3],
+               f3: Callable[[T3], Any],
+               *f_rest: Callable[[Any], Any]) -> Any: ...
+def chain_call(input_value, *f_rest):
+    for function in f_rest:
+        input_value = function(input_value)
+    return input_value
+
+
+class A: ...
+class B: ...
+class C: ...
+class D: ...
+
+def f(x: A) -> A: ...
+def f1(x: A) -> B: ...
+def f2(x: B) -> C: ...
+def f3(x: C) -> D: ...
+
+reveal_type(chain_call(A(), f1, f2))       # E: Revealed type is '__main__.C*'
+reveal_type(chain_call(A(), f1, f2, f3))   # E: Revealed type is 'Any'
+reveal_type(chain_call(A(), f, f, f, f))   # E: Revealed type is '__main__.A'
+[builtins fixtures/list.pyi]
+
+[case testOverloadVarargsSelection]
+from typing import overload, Tuple
+@overload
+def f(x: int) -> Tuple[int]: ...
+@overload
+def f(x: int, y: int) -> Tuple[int, int]: ...
+@overload
+def f(*xs: int) -> Tuple[int, ...]: ...
+def f(*args): pass
+
+i: int
+reveal_type(f(i))           # E: Revealed type is 'Tuple[builtins.int]'
+reveal_type(f(i, i))        # E: Revealed type is 'Tuple[builtins.int, builtins.int]'
+reveal_type(f(i, i, i))     # E: Revealed type is 'builtins.tuple[builtins.int]'
+
+reveal_type(f(*[]))         # E: Revealed type is 'builtins.tuple[builtins.int]'
+reveal_type(f(*[i]))        # E: Revealed type is 'builtins.tuple[builtins.int]'
+reveal_type(f(*[i, i]))     # E: Revealed type is 'builtins.tuple[builtins.int]'
```

Comment:
What happens if there is a tuple instead of list? Ideally `f(*(i, i))` should infer `Tuple[int, int]`. If this is the case, then I would add this call to this test.